src/Nethermind/Nethermind.Synchronization/SnapSync/SnapProvider.cs

```diff
@@ -194,44 +194,42 @@ public AddRangeResult AddStorageRange(StorageRange request, SlotsAndProofs respo
+        public AddRangeResult AddStorageRange(long blockNumber, PathWithAccount pathWithAccount, in ValueHash256 expectedRootHash, in ValueHash256? startingHash, IReadOnlyList<PathWithStorageSlot> slots, IReadOnlyList<byte[]>? proofs = null, ValueHash256? hashLimit = null)
         {
             ITrieStore store = _trieStorePool.Get();
             StorageTree tree = new(store.GetTrieStore(pathWithAccount.Path.ToCommitment()), _logManager);
+            bool canRemoveFurtherLock = false;
+            ProgressTracker.IStorageRangeLock? rangeLock = null;
             try
             {
-                (AddRangeResult result, bool moreChildrenToRight) = SnapProviderHelper.AddStorageRange(tree, blockNumber, startingHash, slots, expectedRootHash, proofs);
+                (AddRangeResult result, bool moreChildrenToRight, rangeLock) = SnapProviderHelper.AddStorageRange(tree, blockNumber, startingHash, slots, expectedRootHash, hashLimit, pathWithAccount.Path, _progressTracker, proofs);
 
                 if (result == AddRangeResult.OK)
                 {
+                    canRemoveFurtherLock = (moreChildrenToRight && slots[^1].Path > hashLimit) || !moreChildrenToRight;
                     if (moreChildrenToRight)
                     {
-                        StorageRange range = new()
-                        {
-                            Accounts = new ArrayPoolList<PathWithAccount>(1) { pathWithAccount },
-                            StartingHash = slots[^1].Path
-                        };
-
-                        _progressTracker.EnqueueStorageRange(range);
+                        _progressTracker.EnqueueStorageRange(pathWithAccount, startingHash, slots[^1].Path, hashLimit);
                     }
                 }
                 else if (result == AddRangeResult.MissingRootHashInProofs)
                 {
                     _logger.Trace($"SNAP - AddStorageRange failed, missing root hash {expectedRootHash} in the proofs, startingHash:{startingHash}");
 
-                    _progressTracker.EnqueueAccountRefresh(pathWithAccount, startingHash);
+                    _progressTracker.EnqueueAccountRefresh(pathWithAccount, startingHash, hashLimit);
                 }
                 else if (result == AddRangeResult.DifferentRootHash)
                 {
                     _logger.Trace($"SNAP - AddStorageRange failed, expected storage root hash:{expectedRootHash} but was {tree.RootHash}, startingHash:{startingHash}");
 
-                    _progressTracker.EnqueueAccountRefresh(pathWithAccount, startingHash);
+                    _progressTracker.EnqueueAccountRefresh(pathWithAccount, startingHash, hashLimit);
                 }
 
                 return result;
             }
             finally
             {
+                rangeLock?.Decrement(pathWithAccount.Path, canRemoveFurtherLock);
```

Comment:
Why do we need to take care of whether a lock should be removed or not? Should it not be the case that when the counter hits 0, we always remove?